### PR TITLE
Implement gateway caching - Closes #710

### LINKED ISSUE #710: Implement gateway caching
### Description

Implement Redis-based request caching.

### Acceptance Criteria

- Responses are stored in Redis
- Add boolean option `ENABLE_REQUEST_CACHING`

### PR DESCRIPTION
**Pull request by nagdahimanshu**
### What was the problem?
This PR resolves #710 

### How was it solved?
- [x] Responses are stored in Redis
- [x] Add boolean option ENABLE_REQUEST_CACHING
- [x] Update docker configuration

### How was it tested?
Local + Jenkins

### CODE REVIEW

File: services/gateway/shared/moleculer-io/index.js [hunk 1]
```diff
@@ -13,10 +13,14 @@ const BluebirdPromise = require('bluebird');
 const {
 	Constants: { JSON_RPC: { INVALID_REQUEST, METHOD_NOT_FOUND, SERVER_ERROR } },
 	Utils,
+	CacheLRU,
```
Comment (sameersubudhi) [inline services/gateway/shared/moleculer-io/index.js:16]:
```suggestion
	CacheRedis,
```

File: services/gateway/shared/moleculer-io/index.js [hunk 2]
```diff
@@ -13,10 +13,14 @@ const BluebirdPromise = require('bluebird');
...
 } = require('lisk-service-framework');
 const config = require('../../config');
 const { BadRequestError } = require('./errors');
 
+const rpcCache = CacheLRU('rpcCache', config.volatileRedis);
```
Comment (sameersubudhi) [inline services/gateway/shared/moleculer-io/index.js:21]:
```suggestion
const rpcCache = CacheRedis('rpcCache', config.volatileRedis);
```

File: services/gateway/shared/moleculer-io/index.js [hunk 3]
```diff
@@ -165,7 +169,18 @@ module.exports = {
...
-				let res = await ctx.call(action, request.params, opts);
+				let res;
+				if (config.rpcCache.enable) {
+					const getResponseFromCache = await rpcCache.get(`${JSON.stringify(request.params)}:${request.method}`);
+					if (getResponseFromCache) {
+						res = JSON.parse(getResponseFromCache);
+					} else {
+						res = await ctx.call(action, request.params, opts);
+						if (res.data.length) rpcCache.set(`${JSON.stringify(request.params)}:${request.method}`, JSON.stringify(res), expireMiliseconds);
+					}
```
Comment (sameersubudhi) [inline services/gateway/shared/moleculer-io/index.js:180]:
```suggestion
					const rpcRequestCacheKey = `${request.method}:${JSON.stringify(request.params)}`;
					const cachedResponse = await rpcCache.get(rpcRequestCacheKey);
					if (cachedResponse) {
						res = JSON.parse(cachedResponse);
					} else {
						res = await ctx.call(action, request.params, opts);
						if (res.data.length) await rpcCache.set(rpcRequestCacheKey, JSON.stringify(res), expireMiliseconds);
					}
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]
Please find the recommended changes below:

File: services/gateway/shared/utils.js [hunk 4]
```diff
@@ -93,9 +93,16 @@ const requireAllJson = apiName => {
 
 const getSwaggerDescription = params => `${params.description}\n RPC => ${params.rpcMethod}`;
 
+const isValidResponse = res => {
+	if (Array.isArray(res.data) && res.data.length) return true;
```
Comment (sameersubudhi) [inline services/gateway/shared/utils.js:97]: An empty array within the `data` is also a valid response. If we want not to cache the empty responses, I'd recommend updating the method name to `isValidNonEmptyResponse` or something similar.

```suggestion
	if (Array.isArray(res.data)) return true;
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]
Except for the mentioned concern, the remaining seem good.

File: services/gateway/shared/utils.js (hunk 4 above)
Comment (nagdahimanshu) [reply services/gateway/shared/utils.js:97]: Updated the method name to `isValidNonEmptyResponse`.

**Review by sameersubudhi — APPROVED** [review]
LGTM 👍

**Review by MichalTuleja — APPROVED** [review]

**Comment by MichalTuleja** [comment]
```
> artillery run benchmark/socket_rpc_v2.yaml

Started phase 0, duration: 15s @ 18:01:04(+0200) 2021-08-31
Report @ 18:01:14(+0200) 2021-08-31
Elapsed time: 10 seconds
  Scenarios launched:  499
  Scenarios completed: 499
  Requests completed:  1143
  Mean response/sec: 114.64
  Response time (msec):
    min: 0
    max: 10.9
    median: 0.1
    p95: 0.8
    p99: 2.4
  Codes:
    0: 1143

Report @ 18:01:20(+0200) 2021-08-31
Elapsed time: 15 seconds
  Scenarios launched:  251
  Scenarios completed: 251
  Requests completed:  529
  Mean response/sec: 96.53
  Response time (msec):
    min: 0
    max: 6.8
    median: 0.1
    p95: 0.4
    p99: 1.4
  Codes:
    0: 529

All virtual users finished
Summary report @ 18:01:20(+0200) 2021-08-31
  Scenarios launched:  750
  Scenarios completed: 750
  Requests completed:  1672
  Mean response/sec: 107.94
  Response time (msec):
    min: 0
    max: 10.9
    median: 0.1
    p95: 0.7
    p99: 2.3
  Scenario counts:
    Accounts: 166 (22.133%)
    Blocks: 203 (27.067%)
    Delegates: 175 (23.333%)
    Transactions: 206 (27.467%)
  Codes:
    0: 1672
```

**Pull request merged**